src/scportrait/pipeline/extraction.py

```diff
@@ -524,206 +509,145 @@ def _extract_classes(self, px_center, arg, return_failed_ids=False):
-                image_data = self.image_data[:, window_y, window_x]
+                # nuclei_mask = self.sdata[self.nucleus_key].data[window_y, window_x].compute()
+                mask = self.seg_masks[mask_ix, window_y, window_x]
             else:
-                image_data = self.image_data[image_index, :, window_y, window_x]
+                # nuclei_mask = self.sdata[self.nucleus_key].data[image_index, window_y, window_x].compute()
+                mask = self.seg_masks[image_index, mask_ix, window_y, window_x]
 
-            image_data = (
-                image_data * masks[-1]
-            )  # always uses the last available mask, in nucleus only seg its the nucleus, if both its the cytosol, if only cytosol its also the cytosol. This always is the mask we want to use to extract the channel information
+            # modify nucleus mask to only contain the nucleus of interest and perform some morphological operations
+            mask = np.where(mask == ids[mask_ix], 1, 0)
+            mask = binary_fill_holes(mask)
+            mask = gaussian(mask, preserve_range=True, sigma=1)
 
-            # this needs to be performed on a per channel basis!
-            images = []
-            for i in range(image_data.shape[0]):
-                ix = self.norm_function(image_data[i])
-                images.append(ix)
+            masks.append(mask)
 
-            inputs = masks + images
-            stack = np.stack(inputs, axis=0).astype(self.DEFAULT_SINGLE_CELL_IMAGE_DTYPE)
+        # get the image data
+        if image_index is None:
+            image_data = self.image_data[:, window_y, window_x]
+        else:
+            image_data = self.image_data[image_index, :, window_y, window_x]
 
-            self._save_cell_info(save_index, cell_id, image_index, label_info, stack)
+        image_data = (
+            image_data * masks[-1]
+        )  # always uses the last available mask, in nucleus only seg its the nucleus, if both its the cytosol, if only cytosol its also the cytosol. This always is the mask we want to use to extract the channel information
 
-            if self.deep_debug:
-                # visualize some cells for debugging purposes
-                if index % 1000 == 0:
-                    print(f"Cell ID: {cell_id} has center at [{_px_center[0]}, {_px_center[1]}]")
+        # this needs to be performed on a per channel basis!
+        images = []
+        for i in range(image_data.shape[0]):
+            ix = self.norm_function(image_data[i])
+            images.append(ix)
 
-                    fig, axs = plt.subplots(1, stack.shape[0], figsize=(2 * stack.shape[0], 2))
-                    for i, img in enumerate(stack):
-                        axs[i].imshow(img, vmin=0, vmax=1)
-                        axs[i].axis("off")
-                    fig.tight_layout()
-                    fig.show()
+        inputs = masks + images
+        stack = np.stack(inputs, axis=0).astype(self.DEFAULT_SINGLE_CELL_IMAGE_DTYPE)
```

Comment:
masks and images same dtype? if true, maybe explain?

Reply:
yes they should have the same dtype. Added an explanation as a comment.